src/pratt.rs

```diff
@@ -0,0 +1,371 @@
+            Associativity::Right => Strength::Strong(self.strength),
+        }
+    }
+
+    /// Get the binding power of this operator with an argument on the right.
+    fn strength_right(&self) -> Strength<T> {
+        match self.associativity {
+            Associativity::Left => Strength::Strong(self.strength),
+            Associativity::Right => Strength::Weak(self.strength),
+        }
+    }
+}
+
+/// Enable Pratt parsing for a binary infix operator.
+pub trait InfixOperator<Expr> {
+    /// The type used to represent operator binding strength.
+    ///
+    /// Unless you have more than 256 operators,
+    /// [`u8`] should be a good choice.
+    type Strength: Copy + Ord;
+
+    /// Get the parsing precedence of this operator.
+    ///
+    /// If an expression has an operator on both sides,
+    /// the one with the greatest strength will
+    /// be built first.
+    ///
+    /// For example, given `x + y * z` where `*` has a greater strength
+    /// than `+` (as usual), the `y` will be combined with the `z` first.
+    /// Next, the combination `(y * z)` will be combined with `x`,
+    /// resulting in `(x + (y * z))`.
+    ///
+    /// If both sides have operators with the same strength,
+    /// then the associativity will determine which side
+    /// will be combined first.
+    ///
+    /// For example, given `x + y + z`;
+    /// left-associativity will result in `((x + y) + z)`,
+    /// while right-associativity will result in `(x + (y + z))`,
+    fn precedence(&self) -> InfixPrecedence<Self::Strength>;
+
+    /// Build an expression for this operator given two arguments.
+    fn build_expression(self, left: Expr, right: Expr) -> Expr;
+}
+
+/// See [`Parser::pratt`].
+#[derive(Copy, Clone)]
+pub struct Pratt<Extra, AtomParser, Expr, OpParser, Op> {
+    pub(crate) parser_op: OpParser,
+    pub(crate) parser_atom: AtomParser,
+    pub(crate) phantom: PhantomData<(Extra, Expr, Op)>,
+}
+
+impl<E, AtomParser, Expr, OpParser, Op> Pratt<E, AtomParser, Expr, OpParser, Op> {
+    fn pratt_parse<'a, M, I>(
+        &self,
+        inp: &mut InputRef<'a, '_, I, E>,
+        min_strength: Option<Strength<Op::Strength>>,
+    ) -> PResult<M, Expr>
+    where
+        I: Input<'a>,
+        E: ParserExtra<'a, I>,
+        AtomParser: Parser<'a, I, Expr, E>,
+        OpParser: Parser<'a, I, Op, E>,
+        Op: InfixOperator<Expr>,
+        M: Mode,
+    {
+        let mut left = self.parser_atom.go::<M>(inp)?;
+        loop {
+            let pre_op = inp.save();
+            let (op, prec) = match self.parser_op.go::<Emit>(inp) {
+                Ok(op) => {
+                    let prec = op.precedence();
+                    if prec.strength_left().is_lt(&min_strength) {
+                        inp.rewind(pre_op);
+                        return Ok(left);
+                    }
+                    (op, prec)
+                }
+                Err(_) => return Ok(left),
```

Comment:
Just a heads up, I think this should be:
```
Err(_) => {
    inp.rewind(pre_op);
    return Ok(left);
}
```
Otherwise, things like `2+3R` where R is any random character will still parse and consume the character, preventing it from being used in future parsers.